src/Disks/IO/AsynchronousReadIndirectBufferFromRemoteFS.cpp

```diff
@@ -171,8 +171,9 @@ bool AsynchronousReadIndirectBufferFromRemoteFS::nextImpl()
     if (!hasPendingDataToRead())
         return false;
 
-    size_t size, offset;
+    chassert(file_offset_of_buffer_end <= impl->getFileSize());
```

Comment:
Should we also check the below condition if that there are no tasks in threadpool 
`chassert(file_offset_of_buffer_end >= impl->getImplementationBufferOffset());`
we had discussed this some time back, not sure if it still makes sense to check it here. what do you think?

Reply:
I added it in commit above and the removed it in last commit https://github.com/ClickHouse/ClickHouse/pull/41231/commits/6a02bdc917435eaa32a39955a8e162475974c68e above, because I thought about this assertion more carefully, and concluded that it must be removed because there is the following situation: we cannot add this assertion before checking prefetch_future, because state which we check in this assertion can be concurrently updated, but if we move this assertion after checking the future, then in the case the future is valid, this assert will actually never be true, because `impl->getImplementationBufferOffset()` will actually be ` file_offset_of_buffer_end <= impl->getImplementationBufferOffset()`, but if we check at the future is not valid, then it must be - `file_offset_of_buffer_end == impl->getImplementationBufferOffset()` in case there was no seek, and if there was seek - we cannot compare them at all, because we could also seek backwards or forwards, so adding the assertion will be incorrect as the offset of impl will not be valid after seek.